### PR TITLE
[@types/got] Added beforeErrorHook type

### PR DESCRIPTION
**Pull request by ivywit**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/sindresorhus/got#hooksbeforeerror
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

Notes:

Declaration of `beforeErrorHook` was missing so when implementing hook in a TypeScript project and error was thrown. This adds the hook's type to allow it's use in a TypeScript project.

### CODE REVIEW

**Review by ikokostya — CHANGES_REQUESTED** [review]
Can you revert unrelated style changes (trailing commas, and indentation)?

**Comment by ivywit** [comment]
@ikokostya definitely! I thought I'd caught all of those, should be fixed now.

File: types/got/index.d.ts [hunk 1]
```diff
@@ -170,6 +172,11 @@ declare namespace got {
      */
     type BeforeRetryHook<Options> = (options: Options, error: GotError, retryCount: number) => any;
 
+    /**
+     * @param error Request or parse error.
+     */
+    type BeforeErrorHook = (error: GotError | ParseError | HTTPError | MaxRedirectsError) => Error | Promise<Error>;
```
Comment (ikokostya) [inline types/got/index.d.ts]: `GotError`  type is union of other errors: 

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/6e067da3948185d3c01fc53fa33a09af1c98039b/types/got/index.d.ts#L351

Therefore, other error types can be omitted.

**Review by ikokostya — APPROVED** [review]

File: types/got/index.d.ts [hunk 2]
```diff
@@ -170,6 +172,11 @@ declare namespace got {
      */
     type BeforeRetryHook<Options> = (options: Options, error: GotError, retryCount: number) => any;
 
+    /**
+     * @param error Request or parse error.
```
Comment (ikokostya) [inline types/got/index.d.ts]: nit: I think tsdoc can be omitted too, because error can be not only request or parse error. Anyway it's obvious.

Comment (ivywit) [reply types/got/index.d.ts]: Yeah, I added it to be consistent with the the types above hook types. I'm happy to remove it though.

**Review by ikokostya — APPROVED** [review]

**Pull request merged**